src/java/voldemort/server/niosocket/AsyncRequestHandler.java

```diff
@@ -187,10 +247,7 @@ protected void write(SelectionKey selectionKey) throws IOException {
 
         // If we don't have anything else to write, that means we're done with
         // the request! So clear the buffers (resizing if necessary).
-        if(outputStream.getBuffer().capacity() >= resizeThreshold)
-            outputStream.setBuffer(ByteBuffer.allocate(socketBufferSize));
-        else
-            outputStream.getBuffer().clear();
+        outputStream.clear();
```

Comment:
The comm buffer issue should be alleviated now. The Admin Buffer is still a mess, and I am working on fixing it, expect a pull request in another month or so.

You are also right about BDB. On read-write on our use cases the overall space savings for the heap size is 4%. But this will let the servers double the number of connections being handled without increasing the memory foot print. The space savings on read only is 20-25% as they run with a very small java heap though.

Your case on growing and dropping back is also exact, the only way to avoid that now is modify the config so that you know your 95th percentile so that growing and shrinking does not happen on each request. The other things you can do is pre-allocate a buffer bool and you can do scatter/gather and return the additional buffers at the end back to the pool. Ideally I believe the connections should not have a buffer. They should be just connections and when they get request they should grab a buffer from the pool. Currently Voldemort can only take a certain number of connections because of the buffer associated with the connections. If we make the buffer association with the connection dynamic the Server can handle few 100,000 connections without issues.